pyomo/core/plugins/transform/logical_to_linear.py

```diff
@@ -0,0 +1,216 @@
+"""Transformation from BooleanVar and LogicalConstraint to Binary and Constraints."""
+from pyomo.common.modeling import unique_component_name
+from pyomo.contrib.fbbt.fbbt import compute_bounds_on_expr
+from pyomo.core import TransformationFactory, BooleanVar, VarList, Binary, LogicalConstraint, Block, ConstraintList, \
+    native_types, BooleanVarList, as_boolean
+from pyomo.core.expr.cnf_walker import to_cnf
+from pyomo.core.expr.logical_expr import AndExpression, OrExpression, NotExpression, AtLeastExpression, \
+    AtMostExpression, ExactlyExpression, special_boolean_atom_types, EqualityExpression, InequalityExpression, \
+    RangedExpression
+from pyomo.core.expr.numvalue import native_logical_types, value
+from pyomo.core.expr.visitor import StreamBasedExpressionVisitor
+from pyomo.core.plugins.transform.hierarchy import IsomorphicTransformation
+from pyomo.core.kernel.component_map import ComponentMap
+from pyomo.gdp import Disjunct
+
+
+@TransformationFactory.register("core.logical_to_linear", doc="Convert logic to linear constraints")
```

Comment:
Given what you suggest, I think the operative question is whether other implementations would use `logical_to_mpec` or whether it would be a flag on a single `logical_to_algebraic` transformation. And I think that's mainly a matter of taste.

There is an open question whether this should be `gdp.logical_to_algebraic` as well. If we move towards more unified expression trees in the future, this transformation is really the "parent" for BM/HR/etc, since the logical expression `Y[1].implies(x == 3)` converts to a disjunction.

I've left support for this out of the current implementation for fear of punting a hornet's nest, and yet it is still an intimidating PR to review.

Reply:
Hmmm, yeah, I like the idea about it being a "parent" to the gdp transformations eventually. But in that case it wouldn't be `logical_to_algebraic` at all. It's logical_to_less_logical--I don't know what you would call it. Maybe `logical_to_disjunctive`? But anyway, based on this I think that the specificity of `logical_to_linear` might actually be a good idea, as there is no reason to assume that other transformations would actually go directly to an algebraic model.